src/search/SearchMessagesScreen.js

```diff
@@ -78,33 +62,46 @@ class SearchMessagesScreen extends PureComponent<Props, State> {
   };
 
   handleQueryChangeInner = async (id: number, query: string) => {
-    let messages: Message[];
-    {
-      // if the promise's construction fails, we let the exception
-      // propagate immediately
-      const networkPromise = this.performQueryRaw(query);
-
-      try {
-        messages = await networkPromise;
-      } finally {
-        /* Succeed or fail, we update our request-state. We discard late
+    const { dispatch } = this.props;
+    let apiResponse;
+    const fetchArgs = {
+      narrow: SEARCH_NARROW(query),
+      anchor: LAST_MESSAGE_ANCHOR,
+      numBefore: 20,
+      numAfter: 0,
+    };
+
+    // if the promise's construction fails, we let the exception
+    // propagate immediately
+    try {
+      apiResponse = await api.getMessages(this.props.auth, fetchArgs);
+    } finally {
+      /* Succeed or fail, we update our request-state. We discard late
            results _even if they are errors_. */
-        if (this.lastIdReceived > id) {
-          // eslint-disable-next-line no-unsafe-finally
-          return;
-        }
-        this.lastIdReceived = id;
+      if (this.lastIdReceived > id) {
+        // eslint-disable-next-line no-unsafe-finally
+        return;
       }
-      /* TODO: if an error makes it through the filter above,
-         should we arrange to display something to the user? */
+      this.lastIdReceived = id;
     }
-    // N.B.: `messages` is now set
+    /* TODO: if an error makes it through the filter above,
+         should we arrange to display something to the user? */
 
     // A query is concluded. Report the message-list.
     this.setState({
-      messages,
+      messages: apiResponse.messages,
       isFetching: this.lastIdSent !== this.lastIdReceived,
     });
+
+    // store the fetched messages in the Redux store.
+    dispatch(
+      messageFetchComplete({
+        ...fetchArgs,
+        messages: apiResponse.messages,
+        foundNewest: apiResponse.found_newest,
+        foundOldest: apiResponse.found_oldest,
+      }),
+    );
```

Comment:
* I think this should go before the `setState`, not after.

  I *think* the difference doesn't currently matter... but the `setState` may trigger a `render` call, and so in principle if that means a `render` call (and a series of further `render` calls on descendant components) before we have the new data stowed away in our Redux state then something could get confused by not finding data about a message.

* This will cause a `MESSAGE_FETCH_COMPLETE` action to be dispatched for this search narrow. Do all our reducers handle that appropriately?

  Tip: searching for `case MESSAGE_FETCH_COMPLETE` is great for finding all the places we act on a given action type in our reducers.

  I think in the `narrows` reducer we probably want to skip search narrows. We're not going to keep them up to date there (as new messages come in as events), and we'd accumulate a bunch of old searches.

  The `caughtUp` and `fetching` reducers should do the same as `narrows`.

  The `messages` reducer is the reason we're doing this PR, and `flags` goes along with it. I think those don't need any adjustment.

Reply:
Changed the reducers.